lib/Solver/Z3Builder.cpp

```diff
@@ -400,16 +401,18 @@ Z3ASTHandle Z3Builder::getInitialArray(const Array *root) {
     array_expr = buildArray(unique_name.c_str(), root->getDomain(),
                             root->getRange());
 
-    if (root->isConstantArray()) {
-      // FIXME: Flush the concrete values into Z3. Ideally we would do this
-      // using assertions, which might be faster, but we need to fix the caching
-      // to work correctly in that case.
-      for (unsigned i = 0, e = root->size; i != e; ++i) {
-        Z3ASTHandle prev = array_expr;
-        array_expr = writeExpr(
-            prev, construct(ConstantExpr::alloc(i, root->getDomain()), 0),
-            construct(root->constantValues[i], 0));
+    if (root->isConstantArray()
+        && constant_array_assertions.count(root) == 0) {
+     std::vector<Z3ASTHandle> array_assertions;
+     for (unsigned i = 0, e = root->size; i != e; ++i) {
+        //construct(= (select i root) root->value[i]) to be asserted in Z3Solver.cpp
+        array_assertions.emplace_back(
+          Z3_mk_eq(ctx,
+                   Z3_mk_select(ctx, array_expr, bvConst32(root->getDomain(), i)),
```

Comment:
Don't use bvconst32. We should not be making assumptions about the number of bits used in the domain for the array.

Reply:
The `bvconst32` has an unfortunate name, the 32 doesn't refer to the type of the constant. That is passed as a width argument (root->getDomain()). The 32 referes to the size of the constant for constructing, so `i` in this case. I don't think we need to support arrays over size of 2 billion, but I can use the 64bit version. 

Ideally, I guess we would want to remove the bvConst32 function, and just have bvConst function that is the bvConst64 one, but I don't think this belongs in this pull request.

Reply:
Ah I forgot about that. The `bvConst32()` function is fine then.